### PR TITLE
Rewrite v-dot using fused multiply-add. Some code cleanup and refuctoring

### PR DESCRIPTION
**Pull request by bpecsek**

### CODE REVIEW

File: lisp/lisprb.lisp [hunk 1]
```diff
@@ -233,11 +221,11 @@
               (sphere-new #.(v 0.0 5.0 -1.0) 4.0 #.(v 1.0 0.0 0.0) nil)
               (sphere-new #.(v 8.0 5.0 -1.0) 2.0 #.(v 0.0 0.0 1.0) nil))))
 
-(defun world-camera (world)
```
Comment (guicho271828) [inline lisp/lisprb.lisp:218]: Using macros for this kind of thing is an anti-pattern. Use inlining instead

**Comment by guicho271828** [comment]
Did you check the validity of the result? Did the ppm file match the output of  crb ?
I chose not to submit a PR for my fused-multiply-add because the result changed (due to single-rounding vs double rounding).
Compare the output with `md5sum`.

File: lisp/lisprb.lisp (hunk 1 above)
Comment (bpecsek) [reply lisp/lisprb.lisp:218]: Fine. Code updated.

**Comment by bpecsek** [comment]
Macros changed to inlined functions

Result is verified. ppm image looks good. FMA can increase accuracy so it should not cause negative effect (random noise is different).
Quite a few codes are using different RNG. All of them are wrong? If I used double-float the result would be wrong?
If I change the seeds, samples, max_depth, the random noise pattern is differently random but it shouldn't imply that the code is wrong.

However, please feel free to verify it yourself too and if you feel that because the random noise is different it is incorrect then it's up to the site maintainer to decide.
And also hope that none of the optimizing compilers are clever enough to pull in some fma.

**Comment by guicho271828** [comment]
Oh, I thought all implementations initialized the random seed to the same value to verify the result.
(double-single is another story.)

**Comment by bpecsek** [comment]
> Oh, I thought all implementations initialized the random seed to the same value to verify the result.
> 
> (double-single is another story.)
> 
> 
Hi Masataro,

Some crystal, elixir, haskell, julia, ocaml, python, ruby and I think rust use different RNGs?

I have also sent you a message on Reddit regarding code speeds on my laptop.

Could you please reply when you have some free time?

Thank You.

Kindest Regards,
Bela

**Comment by niofis** [comment]
Thank you so much for sending updates for this code.

As mentioned in the comments, some implementations are still using different RNGs, but that should not be the case in the future. Code is moving to XorShift RNG for two reasons:

1. To make it easy to check if the result is correct. Since the random values should be the same, the result should be too and correctness is trivial to check (md5, simple diff, etc).
2. Remove the RNG variable. While working on these implementations, I noticed that some languages use their own RNG implementation while others use the one provided by the OS, they vary widely in performance and quality. XorShift is a good tradeoff.

Having said that, the algorithm used for rendering the image is a Monte Carlo approximation so the result should be fine as long as the image looks indistinguishable from the others to the naked eye. If the resulting PPM is different from the one generated by crb (as mentioned by Masataro), then it should still be valid, but I would like you to send the update as an alternative implementation (maybe: lisprb_alt.lisp), to showcase the capabilities of the language not available directly to others.

**Full Disclosure. As you have noticed already, I am not an expert in many of these languages, so let me know if you think this update is good and I'll proceed to merge.**

Again, thank you so much Bela and Masataro for your help!

**Comment by bpecsek** [comment]
Hi there,
The one I PRed is using fused multiply add that is not rounding after the multiply only after the add, therefore, as Masarato pointed out. the result is more accurate then using rounding in each iteration.
However, due to this the resulting image noise in different.
I shell rename the file to lisprb_alt.lisp and resend.

**Comment by bpecsek** [comment]
Could you please make sure that the benchmark is run on a more recent CPU?
The Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz is a very old and SBCL is not running well on it.
Th v4 version and up are all good enough.
I know that you don't have control over it but you can stop the Action and rerun and most likely a better CPU is picked.

**Comment by niofis** [comment]
Thank you so much for your work on this.

I totally agree with your comment about that CPU being very old. I will update the README at some point (soon, I expect) and will look for a more modern VM to run the benchmarks on.

Thank you again, have a great rest of the week!.

**Pull request merged**